_blueprints/work_item_types.adoc

```diff
@@ -0,0 +1,78 @@
+One challenge is to reuse fields where they mean the same thing. When we are looking for all fields with the owner "tmaeder", we don't want to have to search different fields like "scrum.owner" and "waterfall.owner", etc. 
+Another property we would like is for algorithms (services) to be able do work against a large range of work items. For example, a scrum work item might have fields owner, state and sprint. When we decide to plan our project with a kanban board from now on, we would like to reuse the owner and stare fields. In order to implement things like this, services will often have to query for item of a certain type: if a work item does not contain a "owner" fields, for example, it cannot be planned on a scrum board. I propose the following to address these requirements.
+
+== Work Item Type Definition
+=== Field Names
+If two work item types define the same field (by name), they are considered to designate the same field. If the fields have different types, it is an error. In order to allow for wide sharing of fields among services, the base system publishes a list of common field names (system.owner, system.priority, etc.). 
+As we extend the system, the authors of these extensions are encouraged to publish their own well known field names in a common registry (�scrum.sprint�, �scrum.storyPoints�, �scrum.backlogorder�).
+
+=== Basic Types
+The central system will predefine a number of basic work item types like (think "issue", "sprint", "release", "user story", "requirement") in order to enforce a certain common base of fields for a given type of work item. The system will have a hierarchy of predefined types: a root type which contains only very basic fields (last change date, last change user, etc.) and more specific types (like "task"). Furthermore, additional field names and types may be published in order to encourage shared usage of them (think "planning.storyPoints", etc.). 
+
+=== Type Inheritance
+New Types can be added by inheriting from existing types. Fields can be added, but they must conform to the field name rules system-wide. Type inheritance is used to define a special kind of work item that can be used in a new way. 
+ 
+=== Type Extension (Mixins)
+Type mixins should be used to extend a type of work items with new fields. This is to address adding cross cutting concerns to a type of work item. For example, one could add fields for time tracking to all work items of the predefined type "task". All work items inheriting from "task" (bug, story, etc.) would inherit the extended fields.
+
+=== Queries & Duck Typing
+Many times a service will want to query for work items of a given type (for example, a scrum board will want items that have "rank", "owner", "sprint" and "state"). I propose that we use duck typing when evaluating such a query. Concretely, any work item type that includes the same structure as the queried for type will match. 
```

Comment:
You mean, pseudo query lang:

Select wi from workitem
where wi.type in(
   select wit from workitemtype
   where wit.fields contain all ('"rank", "owner", "sprint", "state") 
) 

?